File: Python/Day12.py
# Day 12: The N-Body Problem

import AoC
from typing import List, Dict, Set
import re
import functools
from copy import deepcopy
import numpy as np

def parse_moon(line: str) -> List[int]:
	x = re.findall("[-\d]+", line)
	return list(map(lambda val: (int(val)), x))

def apply_gravity(positions: List[List[int]], velocities: List[List[int]]):
	for a in range(0, len(positions)-1):
		for b in range(a+1, len(positions)):
			for axis in range(0, 3):
				if positions[a][axis] < positions[b][axis]:
					velocities[a][axis] += 1
					velocities[b][axis] -= 1
				elif positions[a][axis] > positions[b][axis]:
					velocities[a][axis] -= 1
					velocities[b][axis] += 1

def apply_velocity(positions: List[List[int]], velocities: List[List[int]]):
	for moon in range(0, len(positions)):
		for axis in range(0, 3):
			positions[moon][axis] += velocities[moon][axis]



def calc_energy(positions: List[List[int]], velocities: List[List[int]]) -> int:
	energies = [0,0,0,0]
	for moon in range(0, len(positions)):
		kin = 0
		pot = 0
		for axis in range(0, 3):
			pot += abs(positions[moon][axis])
			kin += abs(velocities[moon][axis])
		energies[moon] += kin * pot

	print(f'Energies: {energies}')
	total_energy = functools.reduce(lambda a,b: a+b, energies)
	print(f'Total energy: {total_energy}')
	return total_energy

def counted_simulate(positions: List[List[int]], velocities: List[List[int]], n_steps: int):
	for step in range(0, n_steps):
		#print(f'({step}) Positions:  {positions}')
		#print(f'({step}) Velocities: {velocities}')
		apply_gravity(positions, velocities)
		apply_velocity(positions, velocities)

	print(f'(Final) Positions:  {positions}')
	print(f'(Final) Velocities: {velocities}')

	energy = calc_energy(positions, velocities)

# def simulate_to_repeat(positions: List[List[int]], velocities: List[List[int]]):
# 	orig_pos = deepcopy(positions)
# 	orig_vel = deepcopy(velocities)

# 	steps = 0
# 	while True:
# 		steps += 1
# 		if steps % 1000000 == 0:
# 			print(f'Step {steps}')
# 		apply_gravity(positions, velocities)
# 		apply_velocity(positions, velocities)
# 		if orig_pos == positions:
# 			if orig_vel == velocities:
# 				break
	
	# print(f"Universe repeated in {steps} steps")

def find_cycle(track: np.ndarray, start_at: int) -> int:
	# See if the first half of the list equals the second half
	assert len(track) % 2 == 0
	for i in range(start_at, len(track), 2):
		half = int(i / 2)
		front = track[:half]
		back = track[half:half+len(front)]
		if np.array_equal(front, back):
			#print(f'found cycle {front[:6]}') # {front} == {back}')
			return half
	return 0

def analyze_cycles(positions: List[List[int]], velocities: List[List[int]]) -> List[List[int]]:
	# How many cycles to a repeat
	# x, y, z for pos and vel for each moon
	cycles = []
	# Stack of values for tracking cycles
	tracking: List[List[np.ndarray]] = []
	for moon in range(0, len(positions)):
		track:List[np.ndarray] = []
		cycles.append([0]*6)
		for i in range(0,6):
			track.append(np.empty(0, dtype=int))
		tracking.append(track)
	
	analyze_after = 10000
	all_cycles_measured = False
	cycle = 0
	while not all_cycles_measured:
		for moon in range(0, len(positions)):
			for axis in range(0, 3):
				if cycles[moon][axis] == 0:
					tracking[moon][axis] = np.append(tracking[moon][axis], positions[moon][axis])
				if cycles[moon][axis+3] == 0:
					tracking[moon][axis+3] = np.append(tracking[moon][axis+3], velocities[moon][axis])
		cycle += 1
		if cycle % analyze_after == 0:
			print(f'cycle {cycle}. So far: {cycles}')
			# Analyze for cycling. Has to be even number of cycles.
			for moon in range(0, len(positions)):
				for i in range(0, 6):
					if cycles[moon][i] == 0:
						# Returns 0 if not found
						c = find_cycle(tracking[moon][i], (cycle - analyze_after + 2))
						if c > 0:
							print(f'moon {moon} axis {i} cycled after {c} steps.')
							cycles[moon][i] = c
			# Are we done?
			all_cycles_measured = True
			for moon in range(0, len(positions)):
				if 0 in cycles[moon]:
					all_cycles_measured = False
					break
		apply_gravity(positions, velocities)
		apply_velocity(positions, velocities)
	
	return cycles

def main():
	all_input = AoC.read_grouped_input_file('Day12_Input.txt')
	# [0] sample with 179 energy after 10 steps
	# [1] sample with 1940 energy after 100 steps
	# [2] challenge with ? energy after 1000 steps
	input_index = 2
	n_steps = 1000

	moon_positions: List[List[int]] = []
	for line in all_input[input_index]:
		moon_positions.append(parse_moon(line))
	moon_velocities: List[List[int]] = []
	for m in range(0, 4):
		moon_velocities.append([0,0,0])

	# Part 1
	counted_simulate(deepcopy(moon_positions), deepcopy(moon_velocities), n_steps)

	# Part 2
	#simulate_to_repeat(moon_positions, moon_velocities) # too slow
	repeat_cycles_per_moon = analyze_cycles(deepcopy(moon_positions), deepcopy(moon_velocities))
	all_cycles: Set[int] = set()
	for moon in repeat_cycles_per_moon:
		for cycle in moon:
			all_cycles.add(cycle)
	repeat_cycles = sorted(list(all_cycles), reverse=True)
	print(repeat_cycles)

	# When do all repeat_cycles sync?
	max_sync = 0
	jump = repeat_cycles[max_sync]
	ts = jump
	remainders = list(map(lambda x: (ts % x), repeat_cycles))

	while True:
		if remainders[0:max_sync+2] == [0] * (max_sync + 2):
			max_sync += 1
			if (remainders == [0] * len(repeat_cycles)):
				print(f"found future state at time {ts}")
				break
			if max_sync < len(repeat_cycles)-2:
				# Don't want to change the jump for the last digit
				jump *= repeat_cycles[max_sync]
				print(f'jump changed to {jump}')
		ts += jump
		remainders = list(map(lambda cycle: (ts % cycle), repeat_cycles))

if __name__ == '__main__':
	main()

File: Java/AoC2019/src/main/java/ca/biickert/aoc2019/Day12.java
package ca.biickert.aoc2019;

import ca.biickert.aoc2019.spatial.Coord3D;
import ca.biickert.aoc2019.util.InputReader;
import ca.biickert.aoc2019.util.Result;
import ca.biickert.aoc2019.util.Solution;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.regex.Pattern;

/**
 *
 * @author sjb
 */
public class Day12 extends Solution {

    public Day12() {
	super(12, "The N-Body Problem", true);
    }

    @Override
    public Result solve(String filename, int index) {
	var result = super.solve(filename, index);

	List<String> input = InputReader.readGroupedInputFile(filename, index);

	var part1Solution = solvePartOne(input);
	var part2Solution = solvePartTwo(input);

	result = new Result(part1Solution, part2Solution);

	return result;
    }

    private String solvePartOne(List<String> input) {
	var moons = parseMoons(input);
	for (int i = 1; i <= 1000; i++) {
	    simulateStep(moons);
	}

	int totalEnergy = 0;
	for (var moon : moons) {
	    totalEnergy += moon.getKineticEnergy() * moon.getPotentialEnergy();
	}

	return String.valueOf(totalEnergy);
    }

    private String solvePartTwo(List<String> input) {
	var repeat = calculateToRepeat(input);

	return String.valueOf(repeat);
    }

    private long calculateToRepeat(List<String> input) {
	var moons = parseMoons(input);
	var cycles = analyzeCycles(moons);
	System.out.println(cycles);
	Set<Integer> allCycles = new HashSet<>();
	for (var moon : cycles) {
	    for (var cycle : moon) {
		allCycles.add(cycle);
	    }
	}
	List<Integer> repeatCycles = allCycles.stream().sorted(Comparator.reverseOrder()).toList();
	System.out.println(repeatCycles);
	
	int maxSync = 0;
	int jump = repeatCycles.get(maxSync);
	long ts = jump;
	long temp = ts; // To avoid "effectively final" lambda capture error
	List<Long> remainders = repeatCycles.stream().map(rc -> temp % rc).toList();
	
	while (true) {
	    boolean remaindersAreZero = true;
	    for (int r = 0; r <= maxSync + 1; r++) {
		remaindersAreZero = remaindersAreZero && remainders.get(r) == 0;
	    }
	    if (remaindersAreZero) {
		maxSync++;
		if (remainders.stream().distinct().count() == 1) {
		    // ALL remainders zero.
		    System.out.println("found future state at time " + ts);
		    break;
		}
		if (maxSync < repeatCycles.size() - 2) {
		    // Don't change the jump for the last remainder digit
		    jump *= repeatCycles.get(maxSync);
		    System.out.println("Jump changed to " + jump);
		}
	    }
	    
	    ts += jump;
	    long temp1 = ts; // To avoid "effectively final" lambda capture error
	    remainders = repeatCycles.stream().map(rc -> temp1 % rc).toList();
	}
	return ts;
    }

    private List<List<Integer>> analyzeCycles(List<Moon> moons) {
	List<List<Integer>> cycles = new ArrayList<>();
	List<List<List<Integer>>> tracking = new ArrayList<>();
	for (Moon moon : moons) {
	    cycles.add(Arrays.asList(new Integer[]{0, 0, 0, 0, 0, 0}));
	    List<List<Integer>> track = new ArrayList();
	    for (int i = 0; i < 6; i++) {
		track.add(new ArrayList<>());
	    }
	    tracking.add(track);
	}

	final int analyzeAfter = 10000;
	boolean allCyclesMeasured = false;
	int cycle = 0;

	while (allCyclesMeasured == false) {
	    for (int m = 0; m < moons.size(); m++) {
		var moon = moons.get(m);
		for (int axis = 0; axis < 6; axis++) {
		    tracking.get(m).get(axis).add(moon.getAxis(axis));
		}
	    }
	    cycle++;
	    if (cycle % analyzeAfter == 0) {
		// Analyze for cycling. cycle must be even.
		for (int m = 0; m < moons.size(); m++) {
		    for (int axis = 0; axis < 6; axis++) {
			if (cycles.get(m).get(axis) == 0) {
			    // Returns 0 if not found
			    var c = findCycle(tracking.get(m).get(axis), cycle - analyzeAfter + 2);
			    if (c > 0) {
				cycles.get(m).set(axis, c);
			    }
			}
		    }
		}
		allCyclesMeasured = true;
		for (int m = 0; m < moons.size(); m++) {
		    if (cycles.get(m).contains(0)) {
			allCyclesMeasured = false;
			break;
		    }
		}
	    }
	    simulateStep(moons);
	}
	return cycles;
    }

    private int findCycle(List<Integer> track, int start) {
	for (int i = start; i < track.size(); i += 2) {
	    int half = i / 2;
	    List<Integer> front = track.subList(0, half);
	    List<Integer> back = track.subList(half, half + front.size());
	    if (front.equals(back)) {
		return half;
	    }
	}
	return 0;
    }

    private long simulateToRepeat(List<String> input) {
	// Too slow for anything except the first test case
	var unchanged = parseMoons(input);
	var moons = parseMoons(input);

	long steps = 0;
	boolean finished = false;
	while (!finished) {
	    simulateStep(moons);
	    steps++;
	    finished = true;
	    for (var m = 0; m < moons.size(); m++) {
		finished = finished && moons.get(m).equals(unchanged.get(m));
	    }
	    if (steps % 100000 == 0) {
		System.out.println(steps);
	    }
	}
	return steps;
    }

    private void simulateStep(List<Moon> moons) {
	for (int a = 0; a < moons.size() - 1; a++) {
	    for (int b = a + 1; b < moons.size(); b++) {
		moons.get(a).interactWith(moons.get(b));
	    }
	}

	for (var moon : moons) {
	    moon.applyVelocity();
	}
    }

    private List<Moon> parseMoons(List<String> input) {
	List<Moon> moons = new ArrayList<>();
	for (var line : input) {
	    moons.add(new Moon(line));
	}
	return moons;
    }
}

class Moon {

    static final Pattern REGEX = Pattern.compile("x=([\\-\\d]+), y=([\\-\\d]+), z=([\\-\\d]+)");
    Coord3D pos;
    Coord3D vel;

    public Moon(String defn) {
	vel = new Coord3D(0, 0, 0);
	var matcher = REGEX.matcher(defn);
	if (matcher.find()) {
	    int x = Integer.parseInt(matcher.group(1));
	    int y = Integer.parseInt(matcher.group(2));
	    int z = Integer.parseInt(matcher.group(3));
	    pos = new Coord3D(x, y, z);
	} else {
	    System.out.println("regex didn't work on " + defn);
	}
    }

    public void interactWith(Moon other) {
	int dx = compare(pos.getX(), other.pos.getX());
	int dy = compare(pos.getY(), other.pos.getY());
	int dz = compare(pos.getZ(), other.pos.getZ());
	vel.setX(vel.getX() + dx);
	vel.setY(vel.getY() + dy);
	vel.setZ(vel.getZ() + dz);
	other.vel.setX(other.vel.getX() - dx);
	other.vel.setY(other.vel.getY() - dy);
	other.vel.setZ(other.vel.getZ() - dz);
    }

    public void applyVelocity() {
	pos = pos.add(vel);
    }

    public int getKineticEnergy() {
	return vel.manhattanDistanceTo(new Coord3D(0, 0, 0));
    }

    public int getPotentialEnergy() {
	return pos.manhattanDistanceTo(new Coord3D(0, 0, 0));
    }

    public int getAxis(int a) {
	switch (a) {
	    case 0:
		return pos.getX();
	    case 1:
		return pos.getY();
	    case 2:
		return pos.getZ();
	    case 3:
		return vel.getX();
	    case 4:
		return vel.getY();
	    case 5:
		return vel.getZ();
	}
	return 0;
    }

    private int compare(int a, int b) {
	if (a < b) {
	    return 1;
	}
	if (a == b) {
	    return 0;
	}
	return -1; // if (b < a)
    }

    @Override
    public String toString() {
	return "pos: " + pos + " vel: " + vel;
    }

    @Override
    public boolean equals(Object o) {
	if (this == o) {
	    return true;
	}
	if (o == null || this.getClass() != o.getClass()) {
	    return false;
	}
	Moon that = (Moon) o;

	return this.pos.equals(that.pos) && this.vel.equals(that.vel);
    }
}
